README.rst
Introduction
============

.. image:: https://secure.travis-ci.org/pyrenees/plone.scale.png
   :target: http://travis-ci.org/pyrenees/plone.scale

This package contains image scaling logic for use in Zope environments. It
supports Zope 2, grok and other systems build on using the Zope ToolKit (ZTK).

Several design goals were used when writing this package:

- image scaling to any width, height, width&height should be supported
  using both up-scaling and down-scaling. Scaling parameters should never
  be fixed in code. This allows designers to use any image scale they want
  without having to modify python code.

- the result of scaling will be an image along with its new size, not a
  HTML or XHTML tag. We already have excellent tools to generate tags in
  the form of Zope Pagetemplates, Genshi and other template languages that
  are much better suited for this purpose.

In addition several implementation goals were defined:

- image scaling must happen on demand instead of up-front. This reduces
  initial save time and prevents unnecessary scalings from being generated.

- image scaling parameters should not be part of the generated URL. Since
  the number of parameters can change and new parameters may be added in
  the future this would create overly complex URLs and URL parsing.

- no HTML rewriting (such as done by `repoze.bitblt`_) should be required.

- it should be possibly to develop an external storage system which stores
  scaled images externally and returns a URL which bypasses the application
  server. This should be configurable via just a filesystem path & base
  URL.

- minimum number of external dependencies, allowing this package to be
  used in many environments.

- testable without requiring zope.testing. Running `setup.py test` should
  be sufficient.

- URLs for scaled images should have an extension which reflects their
  MIME type. This is facilitates cache (and other front-end services)
  configuration.

.. _repoze.bitblt: http://pypi.python.org/pypi/repoze.bitblt


Usage
=====

The most common way to use *plone.scale* is from a HTML template.
In TAL syntax a typical usage looks like this::

  <img tal:define="scales context/@@image-scaling;
                   thumbnail python:scales.scale('logo', width=64, height=64)"
       tal:attributes="src thumbnail/url;
                       width thumbnail/width;
                       height thumbnail/height" />

This generates a thumbnail of an image field called *logo* with a maximum size
of 64x64 pixels. The dimensions of the resulting image (which might not be
exactly 64x64) are set as attributes on the ``img`` tag to speed up browser
rendering.

If you prefer Genshi syntax and have the ``IImageScaleStorage`` interface
in scope the syntax looks like this::

  <img py:with="thumbnail=IImageScaleStorage(context).scale('logo', width=64, heigh=64)"
       py:attributes="dict(src=thumbnail.url, width=thumbnail.width, height=thumbnail.height" />



CHANGES.rst
Changelog
=========


1.5.1 (unreleased)
------------------

Breaking changes:

- *add item here*

New features:

- *add item here*

Bug fixes:

- Fix tests to pass on Python 3.5
  [datakurre]


1.5.0 (2016-05-18)
------------------

New:

- Use an adapter to lookup the actual factory for scaling.
  Deprecated passing the factory as named parameter along,
  because this had not enough flexibility:
  If addons want to provide alterative methods to scale (i.e. cropping),
  now a specific adapter can perform the work.
  [jensens]

Fixes:

- Minor housekeeping.
  [jensens]


1.4.1 (2016-02-12)
------------------

Fixes:

- Fix KeyError in storage.AnnotationStorage._cleanup when attempting
  to delete the storage for the same key twice.
  [fulv]


1.4 (2015-12-07)
----------------

New:

- Resolve conflicts raised when accessing multiple scales concurrently.
  [gotcha]

- Refactored scale storage.
  [gotcha]


1.3.5 (2015-03-10)
------------------

- PIL thumbnail does not work for magnifying images (when scaling up).
  Use resize instead. [sureshvv]


1.3.4 (2014-09-07)
------------------

- When a scale is outdated, discard all image scales that are more
  than a day older than the context.
  Refs https://dev.plone.org/ticket/13791
  [maurits]

- Make sure deleting items or clearing a complete storage works.
  Deleting one item would often delete a linked second item, which
  made it hard to remove several items at once.
  [maurits]


1.3.3 (2014-01-27)
------------------

- Discard old image scales if item was modified.
  Refs https://dev.plone.org/ticket/13791
  [gforcada]

- Generate Progressive JPEG.
  [kroman0]


1.3.2 (2013-05-23)
------------------

- Added a marker interface for scaled image quality.
  Refs http://dev.plone.org/plone/ticket/13337
  [khink]


1.3.1 (2013-04-06)
------------------

- Cropped images are now centralised vertically as well as horizontally [mattss]


1.3 (2013-01-17)
----------------

- Add MANIFEST.in.
  [WouterVH]

- Break up `scaleImage`, so that its scaling-related parts can be applied
  to instances of `PIL.Image` for further processing.
  [witsch]


1.2.2 - 2010-09-28
------------------

- Re-release to fix bad egg created for 1.2.1.
  Refs http://dev.plone.org/plone/ticket/11154
  [witsch]


1.2.1 - 2010-08-18
------------------

- Convert CMYK to RGB, allowing for web previews of print images.
  [tomster]


1.2 - 2010-07-18
----------------

- Update package metadata.
  [hannosch]


1.1 - 2010-04-20
----------------

- Abort if thumbnail behaviour is requested but either width or height is
  missing. This is nicer than confronting the caller with a PIL exception.
  [wichert]

- Rename the `keep` direction to `thumbnail` to make its behaviour more
  intuitive, but accept `keep` for now.
  [wichert]


1.0 - 2010-04-12
----------------

- Only pull in the uuid distribution in Python versions before 2.5.
  [hannosch]

- Don't declare dependency on PIL.
  [davisagli]


1.0a2 - 2010-04-10
------------------

- Add BSD license text following board decision:
  http://lists.plone.org/pipermail/membership/2009-August/001038.html
  [elro]

- Allow to use PIL's thumbnail algorithm to keep the present aspect ratio.
  [spamsch, witsch]

- Allow to set the quality of the resulting image scales.
  [witsch]

- Refactor storage adapter for image scales to be less dependent on the
  underlying content type.
  [witsch]


1.0a1 - 2009-11-10
------------------

- Initial release
  [wichert]


plone/scale/tests/test_scale.py
from io import BytesIO
from plone.scale.scale import scaleImage
from plone.scale.tests import TEST_DATA_LOCATION
from unittest import TestCase

import os.path
import PIL.Image


PNG = open(os.path.join(TEST_DATA_LOCATION, "logo.png"), mode="rb").read()
GIF = open(os.path.join(TEST_DATA_LOCATION, "logo.gif"), mode="rb").read()
CMYK = open(os.path.join(TEST_DATA_LOCATION, "cmyk.jpg"), mode="rb").read()


class ScalingTests(TestCase):

    def testNewSizeReturned(self):
        (imagedata, format, size) = scaleImage(PNG, 42, 51, "down")
        input = BytesIO(imagedata)
        image = PIL.Image.open(input)
        self.assertEqual(image.size, size)

    def testScaledImageIsJpeg(self):
        self.assertEqual(scaleImage(GIF, 84, 103, "down")[1], "JPEG")

    def testScaledCMYKIsRGB(self):
        (imagedata, format, size) = scaleImage(CMYK, 42, 51, "down")
        input = BytesIO(imagedata)
        image = PIL.Image.open(input)
        self.assertEqual(image.mode, "RGB")

    def XtestScaledPngImageIsPng(self):
        # This test failes because the sample input file has a format of
        # None according to PIL..
        self.assertEqual(scaleImage(PNG, 84, 103, "down")[1], "PNG")

    def testSameSizeDownScale(self):
        self.assertEqual(scaleImage(PNG, 84, 103, "down")[2], (84, 103))

    def testHalfSizeDownScale(self):
        self.assertEqual(scaleImage(PNG, 42, 51, "down")[2], (42, 51))

    def testScaleWithCropDownScale(self):
        self.assertEqual(scaleImage(PNG, 20, 51, "down")[2], (20, 51))

    def testNoStretchingDownScale(self):
        self.assertEqual(scaleImage(PNG, 200, 103, "down")[2], (200, 103))

    def testRestrictWidthOnlyDownScale(self):
        self.assertEqual(scaleImage(PNG, 42, None, "down")[2], (42, 52))

    def testRestrictHeightOnlyDownScale(self):
        self.assertEqual(scaleImage(PNG, None, 51, "down")[2], (42, 51))

    def testSameSizeUpScale(self):
        self.assertEqual(scaleImage(PNG, 84, 103, "up")[2], (84, 103))

    def testDoubleSizeUpScale(self):
        self.assertEqual(scaleImage(PNG, 168, 206, "up")[2], (168, 206))

    def testHalfSizeUpScale(self):
        self.assertEqual(scaleImage(PNG, 42, 51, "up")[2], (42, 51))

    def testNoStretchingUpScale(self):
        self.assertEqual(scaleImage(PNG, 200, 103, "up")[2], (84, 103))

    def testRestrictWidthOnlyUpScale(self):
        self.assertEqual(scaleImage(PNG, 42, None, "up")[2], (42, 52))

    def testRestrictHeightOnlyUpScale(self):
        self.assertEqual(scaleImage(PNG, None, 51, "up")[2], (42, 51))

    def testNoRestrictions(self):
        self.assertRaises(ValueError, scaleImage, PNG, None, None)

    def testKeepAspectRatio(self):
        self.assertEqual(scaleImage(PNG, 80, 80, "keep")[2], (65, 80))

    def testQuality(self):
        img1 = scaleImage(CMYK, 84, 103)[0]
        img2 = scaleImage(CMYK, 84, 103, quality=50)[0]
        img3 = scaleImage(CMYK, 84, 103, quality=20)[0]
        self.assertNotEqual(img1, img2)
        self.assertNotEqual(img1, img3)
        self.failUnless(len(img1) > len(img2) > len(img3))

    def testResultBuffer(self):
        img1 = scaleImage(PNG, 84, 103)[0]
        result = BytesIO()
        img2 = scaleImage(PNG, 84, 103, result=result)[0]
        self.assertEqual(result, img2)      # the return value _is_ the buffer
        self.assertEqual(result.getvalue(), img1)   # but with the same value


def test_suite():
    from unittest import defaultTestLoader
    return defaultTestLoader.loadTestsFromName(__name__)


plone/scale/tests/test_storage.py
# -*- coding: utf-8 -*-
from operator import delitem
from operator import itemgetter
from operator import setitem
from plone.testing import zca
from plone.scale.storage import AnnotationStorage
from unittest import TestCase
import sys
import types


class _DummyContext(object):
    pass


class TestStorage(AnnotationStorage):
    storage = None

    def __init__(self, context, modified=None):
        super(TestStorage, self).__init__(context, modified)
        self.storage = {}


class AnnotationStorageTests(TestCase):

    layer = zca.UNIT_TESTING

    def _provide_dummy_scale_adapter(self, result=True):
        from zope.component import adapter
        from zope.component import provideAdapter
        from zope.interface import implementer
        from plone.scale.interfaces import IImageScaleFactory

        factory = self.factory

        @implementer(IImageScaleFactory)
        @adapter(_DummyContext)
        class DummyISF(object):

            def __init__(self, context):
                self.context = context

            def __call__(self, **parameters):
                if result:
                    return factory()
                return None

        provideAdapter(DummyISF)

    @property
    def storage(self):
        storage = TestStorage(_DummyContext())
        storage.modified = lambda: 42
        return storage

    def factory(self, **kw):
        return 'some data', 'png', (42, 23)

    def testInterface(self):
        from plone.scale.storage import IImageScaleStorage
        storage = self.storage
        self.failUnless(IImageScaleStorage.providedBy(storage))

    def testScaleForNonExistingScaleWithCreationBBB(self):
        storage = self.storage
        scale = storage.scale(factory=self.factory, foo=23, bar=42)
        self.failUnless('uid' in scale)
        self.failUnless('key' in scale)
        self.assertEqual(scale['data'], 'some data')
        self.assertEqual(scale['width'], 42)
        self.assertEqual(scale['height'], 23)
        self.assertEqual(scale['mimetype'], 'image/png')

    def testScaleForNonExistingScaleWithCreation(self):
        self._provide_dummy_scale_adapter()
        storage = self.storage
        scale = storage.scale(foo=23, bar=42)
        self.failUnless('uid' in scale)
        self.failUnless('key' in scale)
        self.assertEqual(scale['data'], 'some data')
        self.assertEqual(scale['width'], 42)

        self.assertEqual(scale['height'], 23)
        self.assertEqual(scale['mimetype'], 'image/png')

    def testScaleForNonExistingScaleWithoutCreationBBB(self):
        storage = self.storage
        scale = storage.scale(foo=23, bar=42)
        self.assertEqual(scale, None)

    def testScaleForNonExistingScaleWithoutCreation(self):
        self._provide_dummy_scale_adapter(result=None)
        storage = self.storage
        scale = storage.scale(foo=23, bar=42)
        self.assertEqual(scale, None)

    def testScaleForExistingScaleBBB(self):
        storage = self.storage
        scale1 = storage.scale(factory=self.factory, foo=23, bar=42)
        scale2 = storage.scale(factory=self.factory, bar=42, foo=23)
        self.failUnless(scale1 is scale2)

    def testScaleForExistingScale(self):
        self._provide_dummy_scale_adapter()
        storage = self.storage
        scale1 = storage.scale(foo=23, bar=42)
        scale2 = storage.scale(bar=42, foo=23)
        self.failUnless(scale1 is scale2)

    def testScaleForSimilarScalesBBB(self):
        storage = self.storage
        scale1 = storage.scale(factory=self.factory, foo=23, bar=42)
        scale2 = storage.scale(factory=self.factory, bar=42, foo=23, hurz='!')
        self.failIf(scale1 is scale2)

    def testScaleForSimilarScales(self):
        self._provide_dummy_scale_adapter()
        storage = self.storage
        scale1 = storage.scale(foo=23, bar=42)
        scale2 = storage.scale(bar=42, foo=23, hurz='!')
        self.failIf(scale1 is scale2)

    def testGetItemBBB(self):
        storage = self.storage
        scale = storage.scale(factory=self.factory, foo=23, bar=42)
        uid = scale['uid']
        scale = storage[uid]
        self.failUnless('uid' in scale)
        self.failUnless('key' in scale)
        self.assertEqual(scale['data'], 'some data')
        self.assertEqual(scale['width'], 42)
        self.assertEqual(scale['height'], 23)
        self.assertEqual(scale['mimetype'], 'image/png')

    def testGetItem(self):
        self._provide_dummy_scale_adapter()
        storage = self.storage
        scale = storage.scale(foo=23, bar=42)
        uid = scale['uid']
        scale = storage[uid]
        self.failUnless('uid' in scale)
        self.failUnless('key' in scale)
        self.assertEqual(scale['data'], 'some data')
        self.assertEqual(scale['width'], 42)
        self.assertEqual(scale['height'], 23)
        self.assertEqual(scale['mimetype'], 'image/png')

    def testGetUnknownItem(self):
        storage = self.storage
        self.assertRaises(KeyError, itemgetter('foo'), storage)

    def testSetItemNotAllowed(self):
        storage = self.storage
        self.assertRaises(RuntimeError, setitem, storage, 'key', None)

    def testIterateWithoutAnnotations(self):
        storage = self.storage
        self.assertEqual(list(storage), [])

    def testIterate(self):
        storage = self.storage
        storage.storage.update(one=None, two=None)
        generator = iter(storage)
        self.assertEqual(set(generator), set(['one', 'two']))

    def testKeys(self):
        storage = self.storage
        storage.storage.update(one=None, two=None)
        if sys.version_info >= (3,):
            self.failUnless(isinstance(storage.keys(), {}.keys().__class__))
        else:
            self.failUnless(isinstance(storage.keys(), list))
        self.assertEqual(set(storage.keys()), set(['one', 'two']))

    def testNegativeHasKey(self):
        storage = self.storage
        self.assertEqual('one' in storage, False)

    def testPositiveHasKey(self):
        storage = self.storage
        storage.storage.update(one=None)
        self.assertEqual('one' in storage, True)

    def testDeleteNonExistingItem(self):
        storage = self.storage
        self.assertRaises(KeyError, delitem, storage, 'foo')

    def testDeleteRemovesItemAndIndexBBB(self):
        storage = self.storage
        scale = storage.scale(factory=self.factory, foo=23, bar=42)
        self.assertEqual(len(storage), 1)
        del storage[scale['uid']]
        self.assertEqual(len(storage), 0)

    def testDeleteRemovesItemAndIndex(self):
        self._provide_dummy_scale_adapter()
        storage = self.storage
        scale = storage.scale(foo=23, bar=42)
        self.assertEqual(len(storage), 1)
        del storage[scale['uid']]
        self.assertEqual(len(storage), 0)

    def testCleanUpOldItemsBBB(self):
        storage = self.storage
        scale_old = storage.scale(factory=self.factory, foo=23, bar=42)
        next_modified = storage.modified() + 1
        storage.modified = lambda: next_modified
        scale_new = storage.scale(factory=self.factory, foo=23, bar=42)
        self.assertEqual(len(storage), 1)
        self.assertEqual(scale_new['uid'] in storage, True)
        self.assertEqual(scale_old['uid'] in storage, False)

        # When modification time is older than a day, too old scales
        # get purged.
        next_modified = storage.modified() + 24 * 60 * 60 * 1000 + 1
        storage.modified = lambda: next_modified
        scale_newer = storage.scale(factory=self.factory, foo=23, bar=42)

        self.assertEqual(scale_newer['uid'] in storage, True)
        self.assertEqual(scale_new['uid'] in storage, False)
        self.assertEqual(scale_old['uid'] in storage, False)
        del storage[scale_newer['uid']]
        self.assertEqual(len(storage), 0)

    def testCleanUpOldItems(self):
        self._provide_dummy_scale_adapter()
        storage = self.storage
        scale_old = storage.scale(foo=23, bar=42)
        next_modified = storage.modified() + 1
        storage.modified = lambda: next_modified
        scale_new = storage.scale(foo=23, bar=42)
        self.assertEqual(len(storage), 1)
        self.assertEqual(scale_new['uid'] in storage, True)
        self.assertEqual(scale_old['uid'] in storage, False)

        # When modification time is older than a day, too old scales
        # get purged.
        next_modified = storage.modified() + 24 * 60 * 60 * 1000 + 1
        storage.modified = lambda: next_modified
        scale_newer = storage.scale(foo=23, bar=42)

        self.assertEqual(scale_newer['uid'] in storage, True)
        self.assertEqual(scale_new['uid'] in storage, False)
        self.assertEqual(scale_old['uid'] in storage, False)
        del storage[scale_newer['uid']]
        self.assertEqual(len(storage), 0)

    def testClearBBB(self):
        storage = self.storage
        storage.scale(factory=self.factory, foo=23, bar=42)
        self.assertEqual(len(storage), 1)
        storage.clear()
        self.assertEqual(len(storage), 0)

    def testClear(self):
        self._provide_dummy_scale_adapter()
        storage = self.storage
        storage.scale(foo=23, bar=42)
        self.assertEqual(len(storage), 1)
        storage.clear()
        self.assertEqual(len(storage), 0)


def test_suite():
    from unittest import defaultTestLoader
    return defaultTestLoader.loadTestsFromName(__name__)


plone/scale/scale.py
# -*- coding: utf-8 -*-
from io import BytesIO

import PIL.Image
import PIL.ImageFile


# Set a larger buffer size. This fixes problems with jpeg decoding.
# See http://mail.python.org/pipermail/image-sig/1999-August/000816.html for
# details.
PIL.ImageFile.MAXBLOCK = 1000000


def scaleImage(image, width=None, height=None, direction="down",
               quality=88, result=None):
    """Scale the given image data to another size and return the result
    as a string or optionally write in to the file-like `result` object.

    The `image` parameter can either be the raw image data (ie a `str`
    instance) or an open file.

    The `quality` parameter can be used to set the quality of the
    resulting image scales.

    The return value is a tuple with the new image, the image format and
    a size-tuple.  Optionally a file-like object can be given as the
    `result` parameter, in which the generated image scale will be stored.

    The `width`, `height`, `direction` parameters will be passed to
    :meth:`scalePILImage`, which performs the actual scaling.
    """
    if isinstance(image, bytes):
        image = BytesIO(image)
    image = PIL.Image.open(image)

    # When we create a new image during scaling we loose the format
    # information, so remember it here.
    format = image.format
    if not format == 'PNG':
        format = 'JPEG'

    image = scalePILImage(image, width, height, direction)

    if result is None:
        result = BytesIO()
        image.save(
            result,
            format,
            quality=quality,
            optimize=True,
            progressive=True)
        result = result.getvalue()
    else:
        image.save(
            result,
            format,
            quality=quality,
            optimize=True,
            progressive=True)
        result.seek(0)

    return result, format, image.size


def scalePILImage(image, width=None, height=None, direction="down"):
    """Scale a PIL image to another size.

    The generated image is a JPEG image, unless the original is a PNG
    image. This is needed to make sure alpha channel information is
    not lost, which JPEG does not support.

    Three different scaling options are supported:

    * `up` scaling scales the smallest dimension up to the required size
      and scrops the other dimension if needed.
    * `down` scaling starts by scaling the largest dimension to the required
      size and scrops the other dimension if needed.
    * `thumbnail` scales to the requested dimensions without cropping. The
      resulting image may have a different size than requested. This option
      requires both width and height to be specified. `keep` is accepted as
      an alternative spelling for this option, but its use is deprecated.

    The `image` parameter must be an instance of the `PIL.Image` class.

    The return value the scaled image in the form of another instance of
    `PIL.Image`.
    """
    if direction == "keep":
        direction = "thumbnail"

    if direction == "thumbnail" and not (width and height):
        raise ValueError(
            "Thumbnailing requires both width and height to be specified")
    elif width is None and height is None:
        raise ValueError("Either width or height need to be given")

    if image.mode == "1":
        # Convert black&white to grayscale
        image = image.convert("L")
    elif image.mode == "P":
        # Convert palette based images to 3x8bit+alpha
        image = image.convert("RGBA")
    elif image.mode == "CMYK":
        # Convert CMYK to RGB, allowing for web previews of print images
        image = image.convert("RGB")

    current_size = image.size
    # Determine scale factor needed to get the right height
    if height is None:
        scale_height = None
    else:
        scale_height = (float(height) / float(current_size[1]))
    if width is None:
        scale_width = None
    else:
        scale_width = (float(width) / float(current_size[0]))

    if scale_height == scale_width or direction == "thumbnail":
        # The original already has the right aspect ratio, so we only need
        # to scale.
        if direction in ("down", "thumbnail"):
            image.thumbnail((width, height), PIL.Image.ANTIALIAS)
        else:
            image = image.resize((width, height), PIL.Image.ANTIALIAS)
    else:
        if direction == "down":
            if scale_height is None or (
                    scale_width is not None and scale_width > scale_height):
                # Width is the smallest dimension (relatively), so scale up
                # to the desired width
                new_width = width
                new_height = int(round(current_size[1] * scale_width))
            else:
                new_height = height
                new_width = int(round(current_size[0] * scale_height))
        else:
            if scale_height is None or (
                    scale_width is not None and scale_width < scale_height):
                # Width is the largest dimension (relatively), so scale up
                # to the desired width
                new_width = width
                new_height = int(round(current_size[1] * scale_width))
            else:
                new_height = height
                new_width = int(round(current_size[0] * scale_height))

        image.draft(image.mode, (new_width, new_height))
        image = image.resize((new_width, new_height), PIL.Image.ANTIALIAS)

        if (width is not None and new_width > width) or (
                height is not None and new_height > height):
            if width is None:
                left = 0
                right = new_width
            else:
                left = int((new_width - width) / 2.0)
                right = left + width
            if height is None:
                top = 0
                bottom = new_height
            else:
                top = int((new_height - height) / 2.0)
                bottom = top + height
            image = image.crop((left, top, right, bottom))

    return image


plone/scale/storage.py
# -*- coding: utf-8 -*-
from persistent.dict import PersistentDict
from plone.scale.interfaces import IImageScaleFactory
try:
    from collections import MutableMapping
except ImportError:
    from UserDict import DictMixin as MutableMapping
from uuid import uuid4
from ZODB.POSException import ConflictError
from zope.annotation import IAnnotations
from zope.interface import implementer
from zope.interface import Interface

import logging
import pprint
import warnings


logger = logging.getLogger('plone.scale')
# Keep old scales around for this amount of milliseconds.
# This is one day:
KEEP_SCALE_MILLIS = 24 * 60 * 60 * 1000


class IImageScaleStorage(Interface):
    """ This is an adapter for image content which can store, retrieve and
        generate image scale data. It provides a dictionary interface to
        existing image scales using the scale id as key. To find or create a
        scale based on its scaling parameters use the :meth:`scale` method. """

    def __init__(context, modified=None):
        """ Adapt the given context item and optionally provide a callable
            to return a representation of the last modification date, which
            can be used to invalidate stored scale data on update. """

    def scale(factory=None, **parameters):
        """ Find image scale data for the given parameters or create it if
            a factory was provided.  The parameters will be passed back to
            the factory method, which is expected to return a tuple
            containing a representation of the actual image scale data (i.e.
            a string or file-like object) as well as the image's format and
            dimensions.  For convenience, this happens to match the return
            value of `scaleImage`, but makes it possible to use different
            storages, i.e. ZODB blobs """

    def __getitem__(uid):
        """ Find image scale data based on its uid. """


class ScalesDict(PersistentDict):

    def raise_conflict(self, saved, new):
        logger.info('Conflict')
        logger.debug('saved\n' + pprint.pformat(saved))
        logger.debug('new\n' + pprint.pformat(new))
        raise ConflictError

    def _p_resolveConflict(self, oldState, savedState, newState):
        logger.debug('Resolve conflict')
        old = oldState['data']
        saved = savedState['data']
        new = newState['data']
        added = []
        modified = []
        deleted = []
        for key, value in new.items():
            if key not in old:
                added.append(key)
            elif value['modified'] != old[key]['modified']:
                modified.append(key)
            # else:
                # unchanged
        for key in old:
            if key not in new:
                deleted.append(key)
        for key in deleted:
            if ((key in saved) and
                    (old[key]['modified'] == saved[key]['modified'])):
                # unchanged by saved, deleted by new
                logger.debug('deleted %s' % repr(key))
                del saved[key]
            else:
                # modified by saved, deleted by new
                self.raise_conflict(saved[key], new[key])
        for key in added:
            if key in saved:
                # added by saved, added by new
                self.raise_conflict(saved[key], new[key])
            else:
                # not in saved, added by new
                logger.debug('added %s' % repr(key))
                saved[key] = new[key]
        for key in modified:
            if key not in saved:
                # deleted by saved, modified by new
                self.raise_conflict(saved[key], new[key])
            elif saved[key]['modified'] != old[key]['modified']:
                # modified by saved, modified by new
                self.raise_conflict(saved[key], new[key])
            else:
                # unchanged in saved, modified by new
                logger.debug('modified %s' % repr(key))
                saved[key] = new[key]
        return dict(data=saved)


@implementer(IImageScaleStorage)
class AnnotationStorage(MutableMapping):
    """ An abstract storage for image scale data using annotations and
        implementing :class:`IImageScaleStorage`. Image data is stored as an
        annotation on the object container, i.e. the image. This is needed
        since not all images are themselves annotatable. """

    def __init__(self, context, modified=None):
        self.context = context
        self.modified = modified

    def _modified_since(self, since):
        if since is None:
            return False
        elif self.modified_time is None:
            return False
        else:
            return self.modified_time > since

    @property
    def modified_time(self):
        if self.modified is not None:
            return self.modified()
        else:
            return None

    def __repr__(self):
        name = self.__class__.__name__
        return '<%s context=%r>' % (name, self.context)

    __str__ = __repr__

    @property
    def storage(self):
        annotations = IAnnotations(self.context)
        scales = annotations.setdefault(
            'plone.scale',
            ScalesDict()
        )
        if not isinstance(scales, ScalesDict):
            # migrate from PersistentDict to ScalesDict
            new_scales = ScalesDict(scales)
            annotations['plone.scale'] = new_scales
            return new_scales
        return scales

    def hash(self, **parameters):
        return tuple(sorted(parameters.items()))

    def get_info_by_hash(self, hash):
        for value in self.storage.values():
            if value['key'] == hash:
                return value

    def scale(self, factory=None, **parameters):
        key = self.hash(**parameters)
        storage = self.storage
        info = self.get_info_by_hash(key)
        if info is not None and self._modified_since(info['modified']):
            del storage[info['uid']]
            # invalidate when the image was updated
            info = None
        elif info is not None:
            return info

        scaling_factory = IImageScaleFactory(self.context, None)

        # BBB/Deprecation handling
        if factory is not None:
            if scaling_factory is not None:
                warnings.warn(
                    'Deprecated usage of factory in plone.scale. '
                    'Factory is passed to plone.scale but also an adapter '
                    'was found. No way to really decide which one to execute.'
                    'To be nice and with a look at backward compatibility the '
                    'passed one is used.',
                    DeprecationWarning
                )
            else:
                warnings.warn(
                    'Deprecated usage of factory in plone.scale. Provide an '
                    'adapter for the factory instead. The kwarg will be '
                    'dropped with plone.scale 3.0',
                    DeprecationWarning
                )
            result = factory(**parameters)
        elif scaling_factory is not None:
            # this is what we want, keep this after deprecaton phase
            result = scaling_factory(**parameters)
        else:
            # adaption error, nor a factory was passed.
            # BBB behavior here is to return None
            # nevertheless we warn!
            warnings.warn(
                'Could not adapt context to IImageScaleFactory nor a '
                'deprecated BBB factory callable was provided.'
                'Assume None return value as it was before.'
            )
            return None

        if result is not None:
            # storage will be modified:
            # good time to also cleanup
            self._cleanup()
            data, format_, dimensions = result
            width, height = dimensions
            uid = str(uuid4())
            info = dict(
                uid=uid,
                data=data,
                width=width,
                height=height,
                mimetype='image/{0}'.format(format_.lower()),
                key=key,
                modified=self.modified_time,
            )
            storage[uid] = info
        return info

    def _cleanup(self):
        storage = self.storage
        modified_time = self.modified_time
        for key, value in storage.items():
            # remove info stored by tuple keys
            # before refactoring
            if isinstance(key, tuple):
                del storage[key]
            # clear cache from scales older than one day
            elif (modified_time and
                    value['modified'] < modified_time - KEEP_SCALE_MILLIS):
                del storage[key]

    def __getitem__(self, uid):
        return self.storage[uid]

    def __setitem__(self, id, scale):
        raise RuntimeError('New scales have to be created via scale()')

    def __delitem__(self, uid):
        del self.storage[uid]

    def __iter__(self):
        return iter(self.storage)

    def keys(self):
        return self.storage.keys()

    def has_key(self, uid):
        return uid in self.storage

    __contains__ = has_key

    def clear(self):
        self.storage.clear()

    def __len__(self):
        return len(tuple(self.keys()))
